README
NAME
    Hopen - Plack based micro web application frameworks.

SYNOPSIS
        use strict;
        use warnings;

        use Hopen;

        get '/' => sub {
            my $c = shift;
            $c->render('index', { name => $c->req->param('name') || 'anonymous' });
        };

        hopen;

         __DATA__

        @@ index
        <h1>Hello [% name %]</h1>
        <form action="/">
          name:<br />
          <input name="name" type="text" />
          <input type="submit" />
        </form>

DESCRIPTION
    Hopen is yet another micro web application framework using Plack,
    Router::Simple, Text::Xslate.

  EXAMPLE
   Using tepmlate in DATA section
        use Hopen;
        get '/' => sub { $_[0]->render('index') };
        hopen;

        __DATA__

        @@ index
        <h1>Hello Hopen</h1>

   Using tepmlate-file
    in app.psgi

        use strict;
        use warnings;

        use Hopen;

        get '/' => sub { $_[0]->render('eg2.tt') };

        hopen;

    in views/eg2.tt

        <h1>Hello Hopen</h1>

    priority: DATA > FILE

   Get params and give args to template
        use strict;
        use warnings;

        use Hopen;

        get '/' => sub {
            my $c = shift;
            $c->render('index', { name => $c->req->param('name') || 'anonymous' });
        };

        hopen;

         __DATA__

        @@ index
        <h1>Hello [% name %]</h1>
        <form action="/">
          name:<br />
          <input name="name" type="text" />
          <input type="submit" />
        </form>

   Handle post request and parse params from url path
        use strict;
        use warnings;

        use Hopen;
        use Text::Xslate qw/uri_escape/;

        get '/' => sub {
            my $c = shift;
            my $name = $c->req->param('name') || 'anonymous';
            $c->redirect('/hello/'.uri_escape($name));
        };

        get '/hello/:name' => sub {
            my ($c, $args) = @_;
            my $name = $args->{name};
            $c->render('hello', { name => $name });
        };

        hopen;

        __DATA__

        @@ hello
        <h1>Hello [% name %]</h1>
        <form action="/">
          name:<br />
          <input name="name" type="text" />
          <input type="submit" />
        </form>

   Make custom response as JSON
        use strict;
        use warnings;

        use Hopen;
        use JSON;

        get '/' => '<a href="/json">download json</a>';

        get '/json' => sub {
             my $c = shift;
             my $json = JSON::to_json({
                 foo => 'FOO',
                 bar => 'BAR',
                 buz => 'BUZ',
             });
             $c->render(
                 [ $json ],
                 {},
                 headers => [ 'Content-Type' => 'application/json' ],
             );
        };

        hopen;

   Using Session
        use strict;
        use warnings;

        use Hopen;
        use Plack::Builder;

        load_plugins('Hopen::Plugin::Session');

        any ['GET', 'POST'], '/' => sub {
             my $c = shift;
             if (uc($c->req->method) eq 'POST') {
                 if ($c->req->param('regen') && $Plack::Middleware::Session::VERSION >= 0.13) {
                     $c->session->options->{change_id}++; # supported P::M::Session >= 0.13
                 }
                 $c->session->set('value', $c->req->param('value'));
                 return $c->redirect('/');
             }
             $c->render('index');
        };

        builder {
            enable 'Session', store => 'File';
            hopen;
        };

        __DATA__

        @@ index
        <form action="/" method="post">
          session value<br/>
          <input name="value" type="text" />
          <label><input type="checkbox" name="regen" value="1"/>regenerate session</label>
          <br/>
          <input type="submit" />
        </form>
        <hr/>
        <dl>
          <dt>session_id</dt>
          <dd>[% c().session.id %]</dd>
          <dt>value</dt>
          <dd>[% c().session.get('value') %]</dd>
        </dl>

   Using Model
    DBI based.

        use strict;
        use warnings;

        use Hopen;
        use File::Temp qw/tempfile/;

        my $conn_info = [
            'dbi:SQLite:'.[ tempfile() ]->[1],
        ];

        { # prepare
            my $dbh = Hopen::DBI->connect(@$conn_info);
            $dbh->do(q{CREATE TABLE IF NOT EXISTS message (id INTEGER PRIMARY KEY, message TEXT)});
        }

        load_plugins('Hopen::Plugin::DBI');

        get '/' => sub {
             my $c = shift;
             my $rows = $c->dbh->selectall_arrayref(
                 q{SELECT message FROM message ORDER BY id DESC},
                 { Slice => {} },
             );
             $c->render('list', { all => $rows });
        };

        post '/edit' => sub {
            my $c = shift;
            if (my $msg = $c->req->param('message')) {
                $c->dbh->do(
                    q{INSERT INTO message (message) VALUES (?)},
                    {},
                    $msg,
                );
            }
            $c->redirect('/list');
        };

        hopen( 'DBI' => $conn_info );

        __DATA__

        @@ list
        <form action="/edit" method="post">
          message<br/>
          <input name="message" type="text" />
          <input type="submit" />
        </form>
        <hr/>
        [%- IF all.size() -%]
        <ul>
          [%- WHILE (item = all.shift()) -%]
          <li>[% item.message %]</li>
          [%- END -%]
        </ul>
        [%- ELSE -%]
        no messages
        [%- END -%]

AUTHOR
    Hayato Imai <hayajo>

SEE ALSO
    Plack, Router::Simple, Text::Xslate, DBI

    Amon2::Lite, Mojolicious::Lite, Dancer

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Hopen

0.04  Mon Aug 15 22:41:09 JST 2011
    - change H::C::render, H::C::redirect and H::C::res_404 return Plack::Response

0.03  Sat Aug 13 23:45:58 JST 2011
    - Clone of Amon2::Lite
    - ほぼ写経なのでオリジナルをご利用ください

0.02  Tue Jul  5 11:52:58 JST 2011
    - added perl version

0.01  Mon Jul  4 15:43:36 JST 2011
    - original version


t/lib/t/Plugin/Hoge.pm
package t::Plugin::Hoge;

use strict;
use warnings;

sub init {
    my ($class, $c, $params) = @_;
    no strict 'refs';
    *{ "$c\::hoge" } = sub { $c->config->{'t::Plugin::Hoge'} || 'hoge' };
}

1;


eg/static/js/facebox/facebox.js
/*
 * Facebox (for jQuery)
 * version: 1.2 (05/05/2008)
 * @requires jQuery v1.2 or later
 *
 * Examples at http://famspam.com/facebox/
 *
 * Licensed under the MIT:
 *   http://www.opensource.org/licenses/mit-license.php
 *
 * Copyright 2007, 2008 Chris Wanstrath [ chris@ozmm.org ]
 *
 * Usage:
 *
 *  jQuery(document).ready(function() {
 *    jQuery('a[rel*=facebox]').facebox()
 *  })
 *
 *  <a href="#terms" rel="facebox">Terms</a>
 *    Loads the #terms div in the box
 *
 *  <a href="terms.html" rel="facebox">Terms</a>
 *    Loads the terms.html page in the box
 *
 *  <a href="terms.png" rel="facebox">Terms</a>
 *    Loads the terms.png image in the box
 *
 *
 *  You can also use it programmatically:
 *
 *    jQuery.facebox('some html')
 *    jQuery.facebox('some html', 'my-groovy-style')
 *
 *  The above will open a facebox with "some html" as the content.
 *
 *    jQuery.facebox(function($) {
 *      $.get('blah.html', function(data) { $.facebox(data) })
 *    })
 *
 *  The above will show a loading screen before the passed function is called,
 *  allowing for a better ajaxy experience.
 *
 *  The facebox function can also display an ajax page, an image, or the contents of a div:
 *
 *    jQuery.facebox({ ajax: 'remote.html' })
 *    jQuery.facebox({ ajax: 'remote.html' }, 'my-groovy-style')
 *    jQuery.facebox({ image: 'stairs.jpg' })
 *    jQuery.facebox({ image: 'stairs.jpg' }, 'my-groovy-style')
 *    jQuery.facebox({ div: '#box' })
 *    jQuery.facebox({ div: '#box' }, 'my-groovy-style')
 *
 *  Want to close the facebox?  Trigger the 'close.facebox' document event:
 *
 *    jQuery(document).trigger('close.facebox')
 *
 *  Facebox also has a bunch of other hooks:
 *
 *    loading.facebox
 *    beforeReveal.facebox
 *    reveal.facebox (aliased as 'afterReveal.facebox')
 *    init.facebox
 *    afterClose.facebox
 *
 *  Simply bind a function to any of these hooks:
 *
 *   $(document).bind('reveal.facebox', function() { ...stuff to do after the facebox and contents are revealed... })
 *
 */
(function($) {
  $.facebox = function(data, klass) {
    $.facebox.loading()

    if (data.ajax) fillFaceboxFromAjax(data.ajax, klass)
    else if (data.image) fillFaceboxFromImage(data.image, klass)
    else if (data.div) fillFaceboxFromHref(data.div, klass)
    else if ($.isFunction(data)) data.call($)
    else $.facebox.reveal(data, klass)
  }

  /*
   * Public, $.facebox methods
   */

  $.extend($.facebox, {
    settings: {
      opacity      : 0.2,
      overlay      : true,
      loadingImage : 'static/js/facebox/loading.gif',
      closeImage   : 'static/js/facebox/closelabel.png',
      imageTypes   : [ 'png', 'jpg', 'jpeg', 'gif' ],
      faceboxHtml  : '\
    <div id="facebox" style="display:none;"> \
      <div class="popup"> \
        <div class="content"> \
        </div> \
        <a href="#" class="close"><img src="static/js/facebox/closelabel.png" title="close" class="close_image" /></a> \
      </div> \
    </div>'
    },

    loading: function() {
      init()
      if ($('#facebox .loading').length == 1) return true
      showOverlay()

      $('#facebox .content').empty()
      $('#facebox .body').children().hide().end().
        append('<div class="loading"><img src="'+$.facebox.settings.loadingImage+'"/></div>')

      $('#facebox').css({
        top:	getPageScroll()[1] + (getPageHeight() / 10),
        left:	$(window).width() / 2 - 205
      }).show()

      $(document).bind('keydown.facebox', function(e) {
        if (e.keyCode == 27) $.facebox.close()
        return true
      })
      $(document).trigger('loading.facebox')
    },

    reveal: function(data, klass) {
      $(document).trigger('beforeReveal.facebox')
      if (klass) $('#facebox .content').addClass(klass)
      $('#facebox .content').append(data)
      $('#facebox .loading').remove()
      $('#facebox .body').children().fadeIn('normal')
      $('#facebox').css('left', $(window).width() / 2 - ($('#facebox .popup').width() / 2))
      $(document).trigger('reveal.facebox').trigger('afterReveal.facebox')
    },

    close: function() {
      $(document).trigger('close.facebox')
      return false
    }
  })

  /*
   * Public, $.fn methods
   */

  $.fn.facebox = function(settings) {
    if ($(this).length == 0) return

    init(settings)

    function clickHandler() {
      $.facebox.loading(true)

      // support for rel="facebox.inline_popup" syntax, to add a class
      // also supports deprecated "facebox[.inline_popup]" syntax
      var klass = this.rel.match(/facebox\[?\.(\w+)\]?/)
      if (klass) klass = klass[1]

      fillFaceboxFromHref(this.href, klass)
      return false
    }

    return this.bind('click.facebox', clickHandler)
  }

  /*
   * Private methods
   */

  // called one time to setup facebox on this page
  function init(settings) {
    if ($.facebox.settings.inited) return true
    else $.facebox.settings.inited = true

    $(document).trigger('init.facebox')
    makeCompatible()

    var imageTypes = $.facebox.settings.imageTypes.join('|')
    $.facebox.settings.imageTypesRegexp = new RegExp('\.(' + imageTypes + ')$', 'i')

    if (settings) $.extend($.facebox.settings, settings)
    $('body').append($.facebox.settings.faceboxHtml)

    var preload = [ new Image(), new Image() ]
    preload[0].src = $.facebox.settings.closeImage
    preload[1].src = $.facebox.settings.loadingImage

    $('#facebox').find('.b:first, .bl').each(function() {
      preload.push(new Image())
      preload.slice(-1).src = $(this).css('background-image').replace(/url\((.+)\)/, '$1')
    })

    $('#facebox .close').click($.facebox.close)
    $('#facebox .close_image').attr('src', $.facebox.settings.closeImage)
  }

  // getPageScroll() by quirksmode.com
  function getPageScroll() {
    var xScroll, yScroll;
    if (self.pageYOffset) {
      yScroll = self.pageYOffset;
      xScroll = self.pageXOffset;
    } else if (document.documentElement && document.documentElement.scrollTop) {	 // Explorer 6 Strict
      yScroll = document.documentElement.scrollTop;
      xScroll = document.documentElement.scrollLeft;
    } else if (document.body) {// all other Explorers
      yScroll = document.body.scrollTop;
      xScroll = document.body.scrollLeft;
    }
    return new Array(xScroll,yScroll)
  }

  // Adapted from getPageSize() by quirksmode.com
  function getPageHeight() {
    var windowHeight
    if (self.innerHeight) {	// all except Explorer
      windowHeight = self.innerHeight;
    } else if (document.documentElement && document.documentElement.clientHeight) { // Explorer 6 Strict Mode
      windowHeight = document.documentElement.clientHeight;
    } else if (document.body) { // other Explorers
      windowHeight = document.body.clientHeight;
    }
    return windowHeight
  }

  // Backwards compatibility
  function makeCompatible() {
    var $s = $.facebox.settings

    $s.loadingImage = $s.loading_image || $s.loadingImage
    $s.closeImage = $s.close_image || $s.closeImage
    $s.imageTypes = $s.image_types || $s.imageTypes
    $s.faceboxHtml = $s.facebox_html || $s.faceboxHtml
  }

  // Figures out what you want to display and displays it
  // formats are:
  //     div: #id
  //   image: blah.extension
  //    ajax: anything else
  function fillFaceboxFromHref(href, klass) {
    // div
    if (href.match(/#/)) {
      var url    = window.location.href.split('#')[0]
      var target = href.replace(url,'')
      if (target == '#') return
      $.facebox.reveal($(target).html(), klass)

    // image
    } else if (href.match($.facebox.settings.imageTypesRegexp)) {
      fillFaceboxFromImage(href, klass)
    // ajax
    } else {
      fillFaceboxFromAjax(href, klass)
    }
  }

  function fillFaceboxFromImage(href, klass) {
    var image = new Image()
    image.onload = function() {
      $.facebox.reveal('<div class="image"><img src="' + image.src + '" /></div>', klass)
    }
    image.src = href
  }

  function fillFaceboxFromAjax(href, klass) {
    $.get(href, function(data) { $.facebox.reveal(data, klass) })
  }

  function skipOverlay() {
    return $.facebox.settings.overlay == false || $.facebox.settings.opacity === null
  }

  function showOverlay() {
    if (skipOverlay()) return

    if ($('#facebox_overlay').length == 0)
      $("body").append('<div id="facebox_overlay" class="facebox_hide"></div>')

    $('#facebox_overlay').hide().addClass("facebox_overlayBG")
      .css('opacity', $.facebox.settings.opacity)
      .click(function() { $(document).trigger('close.facebox') })
      .fadeIn(200)
    return false
  }

  function hideOverlay() {
    if (skipOverlay()) return

    $('#facebox_overlay').fadeOut(200, function(){
      $("#facebox_overlay").removeClass("facebox_overlayBG")
      $("#facebox_overlay").addClass("facebox_hide")
      $("#facebox_overlay").remove()
    })

    return false
  }

  /*
   * Bindings
   */

  $(document).bind('close.facebox', function() {
    $(document).unbind('keydown.facebox')
    $('#facebox').fadeOut(function() {
      $('#facebox .content').removeClass().addClass('content')
      $('#facebox .loading').remove()
      $(document).trigger('afterClose.facebox')
    })
    hideOverlay()
  })

})(jQuery);


lib/Hopen.pm
package Hopen;

use strict;
use warnings;
use 5.008_001;
our $VERSION = '0.04';

use Carp ();
use Data::OptList;
use Data::Section::Simple ();
use File::Basename qw/dirname/;
use File::Spec;
use Plack::Request;
use Router::Simple;
use Text::Xslate;

my $_COUNTER;

{
    our $_CONTEXT; # localize this variable per request
    sub context { $_CONTEXT }
}

sub import {
    strict->import;
    warnings->import;

    no strict "refs";

    my $router = Router::Simple->new();
    my ($caller, $filename) = caller;

    my $base_class = __PACKAGE__.'::__child__'.$_COUNTER++;
    {
        no warnings;
        unshift @{ "$base_class\::ISA" }, (__PACKAGE__, 'Hopen::Context');
        unshift @{ "$caller\::ISA" }, $base_class;
    }

    my $base_dir = File::Spec->rel2abs( dirname($filename) );
    *{ "$caller\::base_dir" } = sub { $base_dir };

    *{ "$caller\::router" } = sub { $router };

    *{ "$caller\::any" } = sub {
        if (@_ == 3) {
            my ($methods, $pattern, $code) = @_;
            $router->connect(
                $pattern,
                { __code => $code },
                { method => [ map { uc $_ } @$methods ] },
            );
        } else {
            my ($pattern, $code) = @_;
            $router->connect( $pattern, { __code => $code } );
        }
    };
    *{ "$caller\::get" } = sub {
        $router->connect($_[0], {__code => $_[1]}, {method => ['GET','HEAD']});
    };
    *{ "$caller\::post" } = sub {
        $router->connect($_[0], {__code => $_[1]}, {method => ['POST']});
    };

    *{ "$caller\::load_plugins" } = sub {
        my @args = @_;
        for my $opt (@{ Data::OptList::mkopt(\@args) }) {
            my ($module, $params) = ($opt->[0], $opt->[1]);
            $module = Plack::Util::load_class($module);
            $module->init($caller, $params);
        }
    };

    *{ "${base_class}\::config" } = sub {
        my $class = shift;
        my %config = (@_ == 1) ? %{$_[0]} : @_;
        no warnings 'redefine';
        *{"$caller\::config"} = sub { \%config };
        \%config;
    };

    *{ "$caller\::hopen" } = sub {
        $caller->config(@_);
        my $app = sub {
            my $env = shift;
            my $req = Plack::Request->new($env);
            my $self = $caller->new($req);
            no warnings 'redefine';
            local $Hopen::_CONTEXT = $self;
            my $res;
            if (my $p = $router->match($env)) {
                my $code = $p->{__code};
               $res = (ref $code eq 'CODE') ? &$code($self, $p)
                                            : $self->render([ $code ]);
                unless (Scalar::Util::blessed($res) && $res->isa('Plack::Response'))  {
                    $res = $self->render([ $res ]);
                }
            } else {
                $res = $self->res_404;
            }
            return $res->finalize;
        };
        $app;
    };

    *{ "${base_class}\::view" } = sub {
        my $vpath = Data::Section::Simple->new($caller)->get_data_section() || +{};
        my $config = $caller->config->{'Text::Xslate'} || +{};
        my $xs_opts = +{
            'syntax' => 'TTerse',
            'module' => [ 'Text::Xslate::Bridge::TT2Like' ],
            'path'   => [ $vpath, File::Spec->catdir($caller->base_dir, 'views') ],
            'function' => {},
            %$config,
        };
        $xs_opts->{function}->{c} = sub { Hopen->context() };
        my $xs = Text::Xslate->new($xs_opts);
        no warnings 'redefine';
        *{ "$caller\::view" } = sub { $xs };
        $xs;
    };

}

package Hopen::Context;

use Encode ();

sub new {
    my ($class, $req) = @_;
    Carp::croak '$req is required'
        unless defined $req && ref($req) eq 'Plack::Request';
    bless {
        encoding => 'utf-8',
        req      => $req,
    }, ref $class || $class;
}

sub view { die "This is abstract method: view" }
sub config { die "This is abstract method: config" }

sub encoding {
    my $self = shift;
    $self->{encoding} = shift if (@_);
    $self->{encoding};
}

sub request { $_[0]->{req} }
sub req     { $_[0]->{req} }

sub create_response { shift->req->new_response(@_) }

sub redirect {
    my ($self, $url, $status) = @_;
    my $res = $self->create_response;
    $res->redirect($url, $status);
    return $res;
}

sub res_404 {
    my $self = shift;
    $self->create_response(
        404,
        { 'Content-Type' => 'text/plain' },
        "not found"
    );
}

sub render {
    my $self = shift;
    my ($tmpl, $vars, %opts) = @_;

    my $body = (ref $tmpl eq 'ARRAY')
        ? $self->view->render_string($tmpl->[0], $vars)
        : $self->view->render($tmpl, $vars);

    $self->create_response(
        $opts{status} || 200,
        $opts{headers} || { 'Content-Type' => 'text/html; charset=' . $self->encoding },
        $self->encode_body($body),
    );
}

sub encode_body {
    my ($self, $body) = @_;
    ($self->encoding) ? Encode::encode($self->encoding, $body) : $body;
}

1;

__END__

=head1 NAME

Hopen - Plack based micro web application frameworks.

=head1 SYNOPSIS

    use strict;
    use warnings;

    use Hopen;

    get '/' => sub {
        my $c = shift;
        $c->render('index', { name => $c->req->param('name') || 'anonymous' });
    };

    hopen;

     __DATA__

    @@ index
    <h1>Hello [% name %]</h1>
    <form action="/">
      name:<br />
      <input name="name" type="text" />
      <input type="submit" />
    </form>

=head1 DESCRIPTION

Hopen is yet another micro web application framework
using Plack, Router::Simple, Text::Xslate.

=head2 EXAMPLE

=head3 Using tepmlate in DATA section

    use Hopen;
    get '/' => sub { $_[0]->render('index') };
    hopen;

    __DATA__

    @@ index
    <h1>Hello Hopen</h1>

=head3 Using tepmlate-file

in app.psgi


    use strict;
    use warnings;

    use Hopen;

    get '/' => sub { $_[0]->render('eg2.tt') };

    hopen;

in views/eg2.tt

    <h1>Hello Hopen</h1>

priority: DATA > FILE

=head3 Get params and give args to template

    use strict;
    use warnings;

    use Hopen;

    get '/' => sub {
        my $c = shift;
        $c->render('index', { name => $c->req->param('name') || 'anonymous' });
    };

    hopen;

     __DATA__

    @@ index
    <h1>Hello [% name %]</h1>
    <form action="/">
      name:<br />
      <input name="name" type="text" />
      <input type="submit" />
    </form>

=head3 Handle post request and parse params from url path

    use strict;
    use warnings;

    use Hopen;
    use Text::Xslate qw/uri_escape/;

    get '/' => sub {
        my $c = shift;
        my $name = $c->req->param('name') || 'anonymous';
        $c->redirect('/hello/'.uri_escape($name));
    };

    get '/hello/:name' => sub {
        my ($c, $args) = @_;
        my $name = $args->{name};
        $c->render('hello', { name => $name });
    };

    hopen;

    __DATA__

    @@ hello
    <h1>Hello [% name %]</h1>
    <form action="/">
      name:<br />
      <input name="name" type="text" />
      <input type="submit" />
    </form>

=head3 Make custom response as JSON


    use strict;
    use warnings;

    use Hopen;
    use JSON;

    get '/' => '<a href="/json">download json</a>';

    get '/json' => sub {
         my $c = shift;
         my $json = JSON::to_json({
             foo => 'FOO',
             bar => 'BAR',
             buz => 'BUZ',
         });
         $c->render(
             [ $json ],
             {},
             headers => [ 'Content-Type' => 'application/json' ],
         );
    };

    hopen;

=head3 Using Session

    use strict;
    use warnings;

    use Hopen;
    use Plack::Builder;

    load_plugins('Hopen::Plugin::Session');

    any ['GET', 'POST'], '/' => sub {
         my $c = shift;
         if (uc($c->req->method) eq 'POST') {
             if ($c->req->param('regen') && $Plack::Middleware::Session::VERSION >= 0.13) {
                 $c->session->options->{change_id}++; # supported P::M::Session >= 0.13
             }
             $c->session->set('value', $c->req->param('value'));
             return $c->redirect('/');
         }
         $c->render('index');
    };

    builder {
        enable 'Session', store => 'File';
        hopen;
    };

    __DATA__

    @@ index
    <form action="/" method="post">
      session value<br/>
      <input name="value" type="text" />
      <label><input type="checkbox" name="regen" value="1"/>regenerate session</label>
      <br/>
      <input type="submit" />
    </form>
    <hr/>
    <dl>
      <dt>session_id</dt>
      <dd>[% c().session.id %]</dd>
      <dt>value</dt>
      <dd>[% c().session.get('value') %]</dd>
    </dl>

=head3 Using Model

DBI based.

    use strict;
    use warnings;

    use Hopen;
    use File::Temp qw/tempfile/;

    my $conn_info = [
        'dbi:SQLite:'.[ tempfile() ]->[1],
    ];

    { # prepare
        my $dbh = Hopen::DBI->connect(@$conn_info);
        $dbh->do(q{CREATE TABLE IF NOT EXISTS message (id INTEGER PRIMARY KEY, message TEXT)});
    }

    load_plugins('Hopen::Plugin::DBI');

    get '/' => sub {
         my $c = shift;
         my $rows = $c->dbh->selectall_arrayref(
             q{SELECT message FROM message ORDER BY id DESC},
             { Slice => {} },
         );
         $c->render('list', { all => $rows });
    };

    post '/edit' => sub {
        my $c = shift;
        if (my $msg = $c->req->param('message')) {
            $c->dbh->do(
                q{INSERT INTO message (message) VALUES (?)},
                {},
                $msg,
            );
        }
        $c->redirect('/list');
    };

    hopen( 'DBI' => $conn_info );

    __DATA__

    @@ list
    <form action="/edit" method="post">
      message<br/>
      <input name="message" type="text" />
      <input type="submit" />
    </form>
    <hr/>
    [%- IF all.size() -%]
    <ul>
      [%- WHILE (item = all.shift()) -%]
      <li>[% item.message %]</li>
      [%- END -%]
    </ul>
    [%- ELSE -%]
    no messages
    [%- END -%]

=head1 AUTHOR

Hayato Imai E<lt>hayajoE<gt>

=head1 SEE ALSO

L<Plack>, L<Router::Simple>, L<Text::Xslate>, L<DBI>

L<Amon2::Lite>, L<Mojolicious::Lite>, L<Dancer>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
